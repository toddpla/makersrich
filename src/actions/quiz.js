import database from '../firebase/firebase'

export const getQuestion = (question) => ({
  type: 'GET_QUESTION',
  question
})

export const startGetQuestion = (uid, random) => {
  return (dispatch) => {
    return database.ref(`players/${uid}/questions`).once('value').then((playerQuestionsSnapshot) => {
      console.log(Object.keys(playerQuestionsSnapshot));
      const playerQuestions = []
      playerQuestionsSnapshot.forEach(playerQuestionsChildSnapshot => {
        playerQuestions.push(playerQuestionsChildSnapshot.key)
      })
      return database.ref('questions').once('value').then((snapshot) => {
        const questions = []
        snapshot.forEach(childSnapshot => {
          questions.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        const availableQuestions = questions.filter(question => {
          return !playerQuestions.includes(question.id)
        })
        if(availableQuestions.length > 0) {
          const index = Math.floor(Math.random() * availableQuestions.length)
          dispatch(getQuestion(availableQuestions[index]))
        }
      })
    })
  }
}

export const startSendResult = ({uid, questionId, result}) => {
  return (dispatch) => {
    return database.ref(`players/${uid}/questions/${questionId}`).update({result})
  }
}
