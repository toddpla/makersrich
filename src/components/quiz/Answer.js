import React from 'react'

export default class Answer extends React.Component {

  handleClick = () => {
    this.props.handleClick(this.props.id)
  }

  render() {
    return (
      <div onClick={this.handleClick}>
        {this.props.answer}
      </div>
    )
  }
}
