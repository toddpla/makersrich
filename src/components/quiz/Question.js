import React from 'react'

export default class Question extends React.Component {
  render() {
    return (
      <div>
        {this.props.question}
      </div>
    )
  }
}
