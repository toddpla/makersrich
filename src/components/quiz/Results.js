import React from 'react'

export default class Results extends React.Component {
  render() {
    return (
      <div>
        4/10 ur shite
      </div>
    )
  }
}
