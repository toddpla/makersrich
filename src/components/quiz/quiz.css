
.quiz-question-box {
  position: absolute;
  top: 150px;
  left: 125px;
  width: 610px;
  height: 75px;
  border-radius: 10px;
  padding-top: 25px;
  font-size: 1.3em;
  background-color: #e1c699
}

.quiz-answer-box {
  border-radius: 10px;
  width: 250px;
  height: 100px;
  overflow-y: scroll;
  padding: 12px 5px;
  font-size: 1.2em;
  background-color: #e1c699
}

.quiz-answer-box:hover {
  background-color: #c19a6b;
  border: 1px solid beige;
  box-shadow: 2px 2px black;
}

.quiz-answer-box:active {
  box-shadow: 4px 4px black
}

.correct-answer {
  background-color: ForestGreen;
}

.correct-answer:hover {
  background-color: ForestGreen;
}

.wrong-answer {
  background-color: FireBrick;
}

.wrong-answer:hover{
  background-color: FireBrick;
}

#answer-0 {
  position: absolute;
  top: 300px;
  left: 125px;
}

#answer-1 {
  position: absolute;
  top: 300px;
  left: 475px;
}

#answer-2 {
  position: absolute;
  top: 450px;
  left: 125px;
}

#answer-3 {
  position: absolute;
  top: 450px;
  left: 475px;
}
