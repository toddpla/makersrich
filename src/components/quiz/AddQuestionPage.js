import React, { Component } from 'react';

import {connect } from 'react-redux'
import { startAddQuestion } from '../../actions/questions'

class AddQuestionPage extends Component {

  handleAddQuestion = (e) => {
    e.preventDefault()
    const question = {
      question: e.target.question.value,
      answers: [
        e.target.answer1.value,
        e.target.answer2.value,
        e.target.answer3.value,
        e.target.answer4.value
      ],
      correctAnswer: e.target.correctAnswer.value
    }
    console.log('here')
    this.props.startAddQuestion(question)
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleAddQuestion}>
          <input type='text' id='question' placeholder='question' />
          <input type='text' id='answer1' placeholder='answer 1' />
          <input type='text' id='answer2' placeholder='answer 2' />
          <input type='text' id='answer3' placeholder='answer 3' />
          <input type='text' id='answer4' placeholder='answer 4' />
          <input type='text' id='correctAnswer'  placeholder='correct answer' />
          <input type='submit' value='submit' />
        </form>
      </div>
    );
  }

}

const mapDispatchToProps = (dispatch) => ({
  startAddQuestion: question => dispatch(startAddQuestion(question))
})

export default connect(undefined, mapDispatchToProps)(AddQuestionPage)
