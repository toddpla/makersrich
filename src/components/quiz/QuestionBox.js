import React from 'react'
import Question from './Question'
import Answer from './Answer'
import { connect } from 'react-redux'

import { loadQuestion, sendResult } from '../../actions/questions'


export class QuestionBox extends React.Component {



  handleClick = (e, index) => {
      var result = false
      if(this.props.answers[index] === this.props.correctAnswer) {
        result = true
      }
      this.props.sendResult(result)
      if(this.props.questionNumber < 4){
        this.props.loadQuestion()
      } else {
        this.props.showResults()
      }

  }

  render() {
    return (
      <div id='question-box-container' style={{display: this.props.visible}}>
        <Question
        question={this.props.question}/>
        <Answer answer={this.props.answers[0]} onClick={ (e) =>  {this.handleClick(e, 0)}} id='answer-1'/>
        <Answer answer={this.props.answers[1]} onClick={ (e) =>  {this.handleClick(e, 1)}} id='answer-2'/>
        <Answer answer={this.props.answers[2]} onClick={ (e) =>  {this.handleClick(e, 2)}} id='answer-3'/>
        <Answer answer={this.props.answers[3]} onClick={ (e) =>  {this.handleClick(e, 3)}} id='answer-4'/>
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => ({
      loadQuestion: () => dispatch(loadQuestion()),
      sendResult: (result) => dispatch(sendResult(result)),
})

export default connect(undefined, mapDispatchToProps)(QuestionBox)
