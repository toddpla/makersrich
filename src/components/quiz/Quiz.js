import React from 'react'
import { connect } from 'react-redux'
import Question from './Question'
import Answer from './Answer'
import { startSendResult, startGetQuestion, clearQuiz } from '../../actions/quiz'
import { startUpdatePlayer, startDebitPlayer, updatePlayer } from '../../actions/auth'
import { QUESTION_PRICE } from '../../constants'

export class Quiz extends React.Component {

  constructor(props) {
    super(props);
    console.log(props);

    if ( this.canAffordQuestion() ) { this.props.startGetQuestion() }

    this.state = {
      questionCount: 0,
    }
  }

  canAffordQuestion = () => {
    return this.props.auth.cash >= QUESTION_PRICE
  }

  sendResultToFirebase = (submission) => {
    if (submission.result === true) {
      this.props.auth.sessionQuestions.push(submission)
      this.props.startSendResult({...submission})
      this.setState({
        correctQuestions: this.state.correctQuestions + 1
      })
      }
    }

  removeQuiz = () => {
    var quizContainer = document.getElementById('quiz-container')
    var quiz = document.getElementById('quiz')
    quizContainer.removeChild(quiz)
  }

  handleClick = (answerIndex) => {
    this.props.startDebitPlayer(QUESTION_PRICE).then(() => {
      const submission = {
        uid: this.props.auth.uid,
        questionId: this.props.quiz.id,
        result: answerIndex.toString() === this.props.quiz.correctAnswer
      }
      this.sendResultToFirebase(submission)

      this.props.clearQuiz()
      if (this.canAffordQuestion()) {
        this.props.startGetQuestion(submission.uid)
      }
      this.setState({
        questionCount: this.state.questionCount + 1
      })
      if (this.state.questionCount === 5) {
        alert('you did it fam!')
      }
    })
  }

  render() {

    return(
      <div id='quiz-container'>
        <h1>Welcome to the quiz!</h1>
          { (Object.keys(this.props.quiz).length !== 0 && this.props.quiz.constructor === Object) ? (
            <div id='quiz'>
              <Question question={this.props.quiz.question}/>
              {this.props.quiz.answers.map((answer, i) => <Answer key={i} id={i} answer={answer} handleClick={this.handleClick} />)}
            </div>
          )
            :
          (
            <div>
              <h2>Come back when you have more money</h2>
            </div>
          )}
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
    quiz: state.quiz,
    auth: state.auth,
    questions: state.questions
})

const mapDispatchToProps = dispatch => ({
  startSendResult: (result) => dispatch(startSendResult(result)),
  startGetQuestion: () => dispatch(startGetQuestion()),
  startUpdatePlayer: (updates) => dispatch(startUpdatePlayer(updates)),
  startDebitPlayer: (amount) => dispatch(startDebitPlayer(amount)),
  clearQuiz: () => dispatch(clearQuiz()),
  updatePlayer: (updates) => dispatch(updatePlayer(updates))
})

export default connect(mapStateToProps, mapDispatchToProps)(Quiz)
