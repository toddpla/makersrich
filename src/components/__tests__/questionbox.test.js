import React from 'react'
import { shallow } from 'enzyme'
import {QuestionBox} from '../quiz/QuestionBox'
import Question from '../quiz/Question'
import Answer from '../quiz/Answer'

describe('QuestionBox', () => {
  it('matches last snapshot', () => {
    const wrapper = shallow(<QuestionBox />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a QuestionBox component', () => {
    const wrapper = shallow(<QuestionBox />)

    expect(wrapper.containsMatchingElement(<Question/>)).toEqual(true);
  })

  it('should render 4 Answer components', () => {
    const wrapper = shallow(<QuestionBox/>)

    expect(wrapper.containsMatchingElement(<Answer />)).toEqual(true);
  })


})
