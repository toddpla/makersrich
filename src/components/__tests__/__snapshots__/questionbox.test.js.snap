// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionBox matches last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionBox
    answers={
      Array [
        1,
        2,
        3,
        4,
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Question />,
        <Answer
          answer={1}
          id="answer-1"
          onClick={[Function]}
        />,
        <Answer
          answer={2}
          id="answer-2"
          onClick={[Function]}
        />,
        <Answer
          answer={3}
          id="answer-3"
          onClick={[Function]}
        />,
        <Answer
          answer={4}
          id="answer-4"
          onClick={[Function]}
        />,
      ],
      "id": "question-box-container",
      "style": Object {
        "display": undefined,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "question": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "answer": 1,
          "id": "answer-1",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "answer": 2,
          "id": "answer-2",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "answer": 3,
          "id": "answer-3",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "answer": 4,
          "id": "answer-4",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Question />,
          <Answer
            answer={1}
            id="answer-1"
            onClick={[Function]}
          />,
          <Answer
            answer={2}
            id="answer-2"
            onClick={[Function]}
          />,
          <Answer
            answer={3}
            id="answer-3"
            onClick={[Function]}
          />,
          <Answer
            answer={4}
            id="answer-4"
            onClick={[Function]}
          />,
        ],
        "id": "question-box-container",
        "style": Object {
          "display": undefined,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "question": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "answer": 1,
            "id": "answer-1",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "answer": 2,
            "id": "answer-2",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "answer": 3,
            "id": "answer-3",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "answer": 4,
            "id": "answer-4",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
